Adds a log of how many network requests were made during dev

orta opened the pull request:
<img width="1035" alt="screen shot 2017-11-09 at 2 11 04 pm" src="https://user-images.githubusercontent.com/49038/32624891-a8931840-c558-11e7-832e-b3e20d100882.png">

Appends to the response data of metaphysics queries to include the number of fetches made during 

```json
{
  "data": {
    [...]
  },
  "extensions": {
    "stagingapi.artsy.net": 42
  }
}
```

This doesn't work in busy environment - it basically expects one graphql request at a time.

I thought about extending our requestID work and using that to map request to keys, which can be done if we want to see this kind of data inside staging also.

orta commented:
/cc @joeyAghion  @anandaroop

craigspaeth commented on lib/apis/fetch/logger.js:
```diff
@@ -0,0 +1,23 @@
+const { URL } = require("url")
+
+export default url => {
+  const cache = global.requestCount
```
It's good to avoid globals at all costs—in this case hoisting this up to the top of the module should suffice, e.g.

````javascript
const { URL } = require("url")

const requestCount = {}

export default url => {
  // ...
}
````

orta replied:
oh yeah, good point, it's all in the same file now

craigspaeth commented:
👍  nice this seems like a good compromise to risking the cost of doing something like this in production. After the global comment is addressed LGTM!

joeyAghion commented:
Thanks for this!

Minor, but `"extensions": { "stagingapi.artsy.net": 42 }` doesn't communicate what this means, or support extending (with other interesting data) down the road. Maybe an additional level of nesting would help, like `"extensions": {"requests": {...}}`?

Also, I recall that many Metaphysics requests attempt to be satisfied from a cache while also queueing a follow-up request to the API to refresh that cached value. Would this count include those?

orta commented:
Updated the naming and removed the global:

![screen shot 2017-11-09 at 3 37 30 pm](https://user-images.githubusercontent.com/49038/32628461-f9271530-c563-11e7-84c8-327ee5e99d39.png)

acjay commented:
How much more expensive would it be to aggregate more details? Like what end points were hit, how many times, and their average latency?

It's a pity there doesn't seem to be a straightforward way to make this concurrency safe.

orta commented:
Yeah, its definitely feasible, the technique can be expanded to pass anything through into the JSON endpoint

Each request is already tied to a request ID, so these stats can get tied to that and then more complicated things can be built on this.

alloy commented:
>> Also, I recall that many Metaphysics requests attempt to be satisfied from a cache while also queueing a follow-up request to the API to refresh that cached value. Would this count include those?
>
> This counts would only include the real HTTP requests made, so a DataLoader freebie wouldn't add to the value.

Seeing as, like @joeyAghion indicates, this could still make a bunch of API requests _after_ the GraphQL query has executed, I wonder if it might be better to count these at the abstraction layer where we know an API request will eventually be made, which is [here](https://github.com/artsy/metaphysics/blob/1a0b095c55cc2ac9b1d3d072ac52e6b9eb83b1ff/lib/loaders/api/loader_without_authentication_factory.js#L38) and [here](https://github.com/artsy/metaphysics/blob/1a0b095c55cc2ac9b1d3d072ac52e6b9eb83b1ff/lib/loaders/api/loader_with_authentication_factory.js#L35)?

orta commented:
> , I wonder if it might be better to count these at the abstraction layer where we know an API request will eventually be made, which is here and here?

I have a branch where I went that way, so that makes sense - it was a lot more code, but the abstraction felt better

orta commented:
Downside is that it can only verify code in a loader, which you could also consider an up-side :P

alloy commented:
Definitely an upside!

orta commented:
This has been updated to use loaders instead:

```graphql
{
  me {
    name
  }
  popular_artists {
    artists{
      name
    }
  }
}
```

now gives

```json
{
  "data": {
    "me": {
      "name": "Orta Therox"
    },
    "popular_artists": {
      "artists": [
        {
          "name": "Pablo Picasso"
        },
       [...]
      ]
    }
  },
  "extensions": {
    "requests": {
      "gravity": {
        "requests": [
          "me",
          "artists/popular"
        ],
        "count": 2
      }
    }
  }
}
```

orta commented:
It now includes the requestID, and requestIDs are now passed on to every service

```json
  "extensions": {
    "requests": {
      "gravity": {
        "requests": [
          "me",
          "artists/popular"
        ],
        "count": 2
      }
    },
    "requestID": "94d0f540-ce2a-11e7-9c94-4944f361f457"
  }
```

orta commented:
@craigspaeth this is good for another look

craigspaeth commented on lib/loaders/api/loader_without_authentication_factory.js:
```diff
@@ -60,6 +63,7 @@ export const apiLoaderWithoutAuthenticationFactory = (api, globalAPIOptions = {}
                 },
                 // Cache miss
                 () => {
+                  logger(apiName)
```
What happens when you use `logger` omitting a `key` here?

craigspaeth commented:
Exciting updates! I have a small comment about the use of `logger` in loader_without_authentication_factory.js that you may want to double check. Otherwise LGTM!

orta commented:
Thanks - great catches, I've tightened the types around this so it won't happen again.

alloy commented on lib/loaders/api/loader_without_authentication_factory.js:
```diff
@@ -60,6 +62,7 @@ export const apiLoaderWithoutAuthenticationFactory = (api, globalAPIOptions = {}
                 },
                 // Cache miss
                 () => {
+                  logger(globalAPIOptions.requestID, apiName, key)
```
@orta This logger should count a fetch regardless of cache hit or miss, because in both cases will it lead to a request.

orta replied:
Agree, I've got an idea for improvements too.
